### DISCUSSION TITLE
Missing segment if subpath has h operator in the middle

### DISCUSSION THREAD

**Discussion by mahardhikapraja** [pymupdf/PyMuPDF#1864] [Q&A] [answered]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
Hi, I try to extracting drawing data from [this pdf](https://github.com/pymupdf/PyMuPDF/files/9279427/test.pdf) but found missing segment at drawing 153.

```python
from fitz import Document


document = Document("path/to/file.pdf")
page = document.load_page(0)
drawing = page.get_drawings()[153]
print(drawing)
```

Got 4 segments from the drawing
```python
{'items': [('l',
   Point(513.6799926757812, 585.3599853515625),
   Point(440.0, 585.3599853515625)),
  ('l',
   Point(440.0, 585.3599853515625),
   Point(513.6799926757812, 587.9400024414062)),
  ('l',
   Point(440.0, 585.3599853515625),
   Point(513.6799926757812, 587.9400024414062)),
  ('l',
   Point(513.6799926757812, 587.9400024414062),
   Point(437.41998291015625, 587.9400024414062))],
 'closePath': False,
 'type': 'fs',
 'stroke_opacity': 1.0,
 'color': (0.3019999861717224, 0.3019999861717224, 0.3019999861717224),
 'width': 0.0,
 'lineCap': (1, 1, 1),
 'lineJoin': 0.0,
 'dashes': '[] 0',
 'rect': Rect(437.41998291015625, 585.3599853515625, 513.6799926757812, 587.9400024414062),
 'seqno': 154,
 'even_odd': False,
 'fill_opacity': 1.0,
 'fill': (0.3019999861717224, 0.3019999861717224, 0.3019999861717224)}
```
The page content that related to the drawing
```python
page.clean_contents()
splitted_contents = page.read_contents().split(b"\n")

# list of painting operators
painting_operators = [b"S", b"s", b"f", b"F", b"f*", b"B", b"B*", b"b", b"b*", b"n"]
painting_counter = -1

seqno = drawing['seqno']


for i, data in enumerate(splitted_contents):
    try:
        values, cmd = data.rsplit(b" ", 1)
    except ValueError:
        values = None
        cmd = data

    if cmd in painting_operators:
        painting_counter += 1
        if painting_counter == seqno:
            break
else:
    raise Exception("TARGER NOT FOUND")

print(splitted_contents[i - 20: i + 1])
```
```python
[
 b'0 G',
 b'S',  # <<< seqno 153
 b'Q',
 b'EMC',
 b'q',
 b'/OC /MC2 BDC',
 b'1 0 0 1 8328 4044 cm',
 b'0 j',
 b'0 w',
 b'0 0 m',  # <<< P1 (0, 0), subpath started
 b'-1228 0 l',  # <<< P2
 b'0 -43 l',  # <<< P3
 b'h',  # <<<  P4 (0, 0), subpath closed
 b'-1228 0 m',  # <<< P5 (-1228, 0), subpath started
 b'0 -43 l',  # <<< P6
 b'-1271 -43 l',  # <<< P7
 b'h',  # <<< P8 (-1228, 0), subpath closed
 b'h',  # <<< P9 (-1228, 0), should be do nothing because subpath already closed
 b'.302 .302 .302 rg',
 b'.302 .302 .302 RG',
 b'B'  # <<< seqno 154
]
```

## To Reproduce (mandatory)
It's been explained above.

## Expected behavior (optional)
6 segments expected in the drawing items
```python
[
 (P1, P2),
 (P2, P3),
 (P3, P4),
 (P5, P6),
 (P6, P7),
 (P7, P8)
]
```

## Your configuration (mandatory)
```python
import sys, fitz
print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)
```
```
3.9.9 (tags/v3.9.9:ccb0e6a, Nov 15 2021, 18:08:50) [MSC v.1929 64 bit (AMD64)] 
 win32 
 
PyMuPDF 1.19.6: Python bindings for the MuPDF 1.19.0 library.
Version date: 2022-03-03 00:00:01.
Built for Python 3.9 on win32 (64-bit).
```

**Comment by JorjMcKie** [comment, marked as answer]
I don't understand your argument:
In `page.get_drawings()` I am not reading the page `/Contents`, but I am accessing general MuPDF interface, which also works for all document types supported. MuPDF does all the abstractions from how whatever document type is doing its rendering.
So cannot derive the contents line number from the "seqno" key in a drawing.

But independently from this, I am counting like that:
There are only the following 4 points:
p1 = (0,0)
p2 = (-1228, 0)
p3 = (0, -43)
p4 = (-1271, -43)

> your p6 equals your p3, and your p5 equals your p2.

And the lines being drawn are (p1, p2), (p2, p3), (p2, p3) **_(again!)_**, (p3, p4). So 4 lines should be drawn, where 2 of them are identical.

**Comment by mahardhikapraja** [comment]
Sorry for my English and thank you for your response.
Is there a way to find out which line of content is related to a drawing?

**Comment by JorjMcKie** [comment]
I think, we can close this as an issue / bug and convert it to a post in the "Discussions" section.
Please see my reaction there.

**Comment by JorjMcKie** [comment]
> Is there a way to find out which line of content is related to a drawing?

No there is not. Except of course you want to (semi-) manually dig your way through the `/Contents` object, re-compute any matrix instruction you would come across, observing the stacking operations of the graphics state (`"q"`/ `"Q"`), etc., etc.
Please be aware that you would some time approach rebuilding a PDF viewer - you probably do not want to do that, don't you.
There also is a MuPDF utility command that produces related output in XML format btw: `mutool trace ...`.
It does a similar thing as `.get_drawings()` and specifically also uses the same low-level MuPDF API as my method.
Differences are that it
* more dilligently observes nesting levels of paint operations
* also produces output like you can get from `page.get_texttrace()`
* outputs **_two separate items_** from objects that are stroked **_and_** filled, whereas `.get_drawings()` combines those cases again in one item with the "fs" type
* does not recognize higher level objects like rects and quads - which `.get_drawings()` does

**Reply by mahardhikapraja** [reply]
Is that possible to accessing mupdf API through PyMuPDF?

**Reply by JorjMcKie** [reply]
> Is that possible to accessing mupdf API through PyMuPDF?

What does that mean? Which MuPDF API? PyMuPDF is a MuPDF binding - which means it already does access the MuPDF API.

**Reply by mahardhikapraja** [reply]
I mean the low level MuPDF API that used by `mutool trace ...` command to collect drawing data.

**Reply by mahardhikapraja** [reply]
I was tried it and the result is correct.

**Reply by JorjMcKie** [reply]
`page.get_drawings()` accesses the MuPDF API. This is C code contained in file `helper-devices.i`
Then you can also redirect `mutool trace` output to a file and inspect the result as XML data with e.g. lxml.

**Reply by mahardhikapraja** [reply]
I will give it a try. Thank you for your advice
